### PR TITLE
Azure OpenAI: add dedicated streaming tool call type (w/index)

### PR DESCRIPTION
**Pull request by trrwilson**
This change addresses usability challenges with streaming tool calls.

### Features Added

**Updates for using streaming tool calls:**

- A new .NET-specific `StreamingToolCallUpdate` type has been added to better represent streaming tool call updates
  when using chat tools.
  - This new type includes an explicit `ToolCallIndex` property, reflecting `index` in the REST schema, to allow
    resilient deserialization of parallel function tool calling.
- A convenience constructor has been added for `ChatRequestAssistantMessage` that can automatically populate from a prior
  `ChatResponseMessage` when using non-streaming chat completions.
- A public constructor has been added for `ChatCompletionsFunctionToolCall` to allow more intuitive reconstruction of
  `ChatCompletionsToolCall` instances for use in `ChatRequestAssistantMessage` instances made from streaming responses.

### Breaking Changes

- The type of `ToolCallUpdate` on `StreamingChatCompletionsUpdate` has been changed from the non-streaming
  `ChatCompletionsToolCall` to the new `StreamingToolCallUpdate` type. The conversion is straightforward:
  - `ToolCallUpdate.Id` remains unchanged.
  - Instead of casting `ToolCallUpdate` to `ChatCompletionsFunctionToolCall`, cast it to `StreamingToolCallUpdate`.
  - Update cast instance use of `functionToolCallUpdate.Arguments` to accumulate `functionToolCallUpdate.ArgumentsUpdate`.

**PR note:** It's possible to work around the lack of this change in the interim: use a `ChatCompletionsFunctionToolCall` instance from the old/current `StreamingChatCompletionsUpdate` and replace its arguments value with the accumulated arguments value from all such instances.

Distinguishing between tool calls can also be inferred without `ToolCallIndex` under the assumption that a new `Id`/`Name` denotes a new function being called. This appears to currently be the case but the use of the new `ToolCallIndex` field in this PR is highly preferred.

### CODE REVIEW

**Review by joseharriaga — APPROVED** [review]

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletions/ChatCompletionsFunctionToolCall.cs [hunk 1]
```diff
@@ -30,6 +30,23 @@ public string Arguments
...
+    /// <param name="arguments"> The arguments to the function that is called by the tool. </param>
+    /// <remarks>
+    /// This constructor is intended for use with <see cref="ChatRequestAssistantMessage"/> when constructing request
+    /// messages for conversation history from accumulated streaming tool call updates.
+    /// </remarks>
+    public ChatCompletionsFunctionToolCall(string id, string name, string arguments) : base(id)
+    {
+        Argument.AssertNotNull(name, nameof(name));
+        Argument.AssertNotNull(arguments, nameof(arguments));
+        Function = new(name, arguments);
```
Comment (stephentoub) [inline sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletions/ChatCompletionsFunctionToolCall.cs:47]: Doesn't this also need to set Type = "function"?

File: sdk/openai/Azure.AI.OpenAI/README.md [hunk 2]
```diff
@@ -301,26 +301,70 @@ calls, and the tool messages that resolved each of those tools -- when making a
...
+        if (functionToolCallUpdate.Name != null)
+        {
+            functionNamesByIndex[functionToolCallUpdate.ToolCallIndex] = functionToolCallUpdate.Name;
+        }
+        if (functionToolCallUpdate.ArgumentsUpdate != null)
+        {
+            StringBuilder argumentsBuilder
+                = functionArgmentBuildersByIndex.TryGetValue(
+                    functionToolCallUpdate.ToolCallIndex,
+                    out StringBuilder existingBuilder) ? existingBuilder : new StringBuilder();
```
Comment (stephentoub) [inline sdk/openai/Azure.AI.OpenAI/README.md:347]: The builder created in the else branch is never stored back into the dictionary. Doesn't that mean all arguments will actually end up being lost?

File: sdk/openai/Azure.AI.OpenAI/README.md [hunk 3]
```diff
@@ -301,26 +301,70 @@ calls, and the tool messages that resolved each of those tools -- when making a
...
+
+ChatRequestAssistantMessage assistantHistoryMessage = new(contentBuilder.ToString());
+foreach (KeyValuePair<int, string> indexIdPair in toolCallIdsByIndex)
+{
+    assistantHistoryMessage.ToolCalls.Add(new ChatCompletionsFunctionToolCall(
+        id: indexIdPair.Value,
+        functionNamesByIndex[indexIdPair.Key],
+        functionArgmentBuildersByIndex[indexIdPair.Key].ToString()));
+}
+chatCompletionsOptions.Messages.Add(assistantHistoryMessage);
```
Comment (stephentoub) [inline sdk/openai/Azure.AI.OpenAI/README.md:365]: Does this work without https://github.com/Azure/azure-sdk-for-net/pull/40704/files#r1429009644 ?

File: sdk/openai/Azure.AI.OpenAI/README.md [hunk 4]
```diff
@@ -301,26 +301,70 @@ calls, and the tool messages that resolved each of those tools -- when making a
...
+
+When using tool calls with streaming responses, accumulate tool call details much like you'd accumulate the other
+portions of streamed choices, in this case using the accumulated `StreamingToolCallUpdate` data to instantiate new
+tool call messages for assistant message history. Note that the model will ignore `ChoiceCount` when providing tools
+and that all streamed responses should map to a single, common choice index in the range of `[0..(ChoiceCount - 1)]`.
+
+```C# Snippet:ChatTools:StreamingChatTools
+Dictionary<int, string> toolCallIdsByIndex = new();
+Dictionary<int, string> functionNamesByIndex = new();
+Dictionary<int, StringBuilder> functionArgmentBuildersByIndex = new();
```
Comment (stephentoub) [inline sdk/openai/Azure.AI.OpenAI/README.md:326]: Typo: Argment => Argument

**Pull request merged**